src/librustc_mir/interpret/place.rs

```diff
@@ -20,6 +20,47 @@ use super::{
     RawConst, Scalar, ScalarMaybeUndef,
 };
 
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, HashStable)]
+/// Information required for the sound usage of a `MemPlace`.
+pub enum MemPlaceMeta<Tag = (), Id = AllocId> {
+    /// The unsized payload (e.g. length for slices or vtable pointer for trait objects).
+    Unsized(Scalar<Tag, Id>),
```

Comment:
There is a potential confusion here because some unsized types don't have metadata -- namely, `extern type`. Maybe `DynSized` would be a better name?

Reply:
I don't know about any difference between the meaning of *unsized* and *dynsized*.

Doesn't `extern type` also have metdata but it's a zst?

Reply:
Dynsized means that dynamically, we can determine the size. That's true for `dyn Trait` but not for `extern type`.

> Doesn't extern type also have metdata but it's a zst?

Well... does it? I might as well argue that `&i32` has metadata but it's a ZST.

Reply:
> That's true for dyn Trait but not for extern type.

Well, it's still up for discussion whether `size_of_val` should work for extern types. Maybe let's stay with `Unsized` and mention `extern type` in the docs?

Reply:
The discussion is only whether it should panic or be prevent to be called through some kind of bound, isn't it? AFAIK the proposal to return a bogus value never got real traction (and rightly so).

Reply:
no, not bogus values. What i mean is that e.g. `CStr` can't become an extern type without us adding some scheme that allows extern types to work with `size_of_val`. Right now we have `struct CStr([u8]);` which works as expected, but if we moved to `struct CStr(ExternCStr);` we'd be doing a breaking change because `size_of_val` would now panic.

Reply:
That sounds to me like the plans for custom DST (dynamically sized types). That's far off, but I see.

Maybe just call the variant `Meta` or so then?